### PR TITLE
Add email notify Message-Id Header

### PR DESCRIPTION
**Pull request by n33pm**
## Problem
Rspamd - `MISSING_MID (2.5)`

Reference: [jordan-wright/email](https://github.com/jordan-wright/email/blob/041e0cec78b06dd35f6431f4fa92f2b29fa30d67/email.go#L612)

### CODE REVIEW

**Comment by simonpasquier** [comment]
Hmm, we can't copy code from https://github.com/jordan-wright without keeping the copyright and the permission notice. I'd prefer that we implement our own id generator and `<timestamp.random_number@hostname>`  is probably enough.

File: notify/email/email.go [hunk 1]
```diff
@@ -239,6 +243,11 @@ func (n *Email) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
 	multipartBuffer := &bytes.Buffer{}
 	multipartWriter := multipart.NewWriter(multipartBuffer)
 
+	id, err := generateMessageID()
```
Comment (simonpasquier) [inline notify/email/email.go]: For sanity, you need to make sure that the `Message-Id` header doesn't already exist.

Comment (n33pm) [reply notify/email/email.go]: done

File: notify/email/email.go [hunk 2]
```diff
@@ -236,6 +243,13 @@ func (n *Email) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
 		fmt.Fprintf(buffer, "%s: %s\r\n", header, mime.QEncoding.Encode("utf-8", value))
 	}
 
+	if reHeaderMessageId.Find(buffer.Bytes()) == nil {
```
Comment (simonpasquier) [inline notify/email/email.go]: I'd rather check that `n.conf.Headers` doesn't contain the `message-id` key.

File: notify/email/email.go [hunk 3]
```diff
@@ -16,15 +16,20 @@ package email
 import (
 	"bytes"
 	"context"
+	"crypto/rand"
```
Comment (simonpasquier) [inline notify/email/email.go]: I think that `math/rand` is enough for what we need here. I would add a `generator *rand.Rand` field to the `Email` struct as well as a `hostname string` field (so `os.Hostname()` is called only once). Then `generateMessageID()` would become a method of `*Email`.

File: notify/email/email.go (hunk 2 above)
Comment (n33pm) [reply notify/email/email.go]: done

File: notify/email/email.go (hunk 3 above)
Comment (n33pm) [reply notify/email/email.go]: - `generateMessageID()` is not needed anymore
- add `os.Hostname()` to `func New`

**Comment by simonpasquier** [comment]
LGTM. I've added a small commit on top to cover the change in the unit tests.

**Comment by n33pm** [comment]
> LGTM. I've added a small commit on top to cover the change in the unit tests.

thx..

**Pull request merged**